styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java

```diff
@@ -117,17 +117,21 @@ public Void created(WorkflowInstance workflowInstance, String executionId, Strin
     }
 
     @Override
-    public Void submit(WorkflowInstance workflowInstance, ExecutionDescription executionDescription) {
+    public Void submit(WorkflowInstance workflowInstance, ExecutionDescription executionDescription,
+        String executionId) {
       currWorkflowInstance = workflowInstance;
       currDockerImg = executionDescription.dockerImage();
+      currExecutionId = executionId;
 
       return null;
     }
 
     @Override
     public Void submitted(WorkflowInstance workflowInstance, String executionId) {
       currWorkflowInstance = workflowInstance;
-      currExecutionId = executionId;
+      if (currExecutionId == null) {
```

Comment:
Changing execution id's is actually expected in some circumstances. E.g. a wfi that is retrying.